// react-responsive-ui/commonjs/ActivityIndicator.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; // http://codepen.io/jczimm/pen/vEBpoL

exports.default = ActivityIndicator;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var radius = 20;
var padding = 0.25; // in `radius`es

var svg_circe_center = radius * (1 + padding);
var svg_canvas_dimensions = '0 0 ' + svg_circe_center * 2 + ' ' + svg_circe_center * 2;
// Whatever it is
// https://developer.mozilla.org/docs/Web/SVG/Attribute/stroke-miterlimit
var svg_stroke_miter_limit = radius / 2;

function ActivityIndicator(props) {
	var className = props.className;


	return _react2.default.createElement(
		'div',
		_extends({}, props, {
			className: (0, _classnames2.default)('rrui__activity-indicator', className) }),
		_react2.default.createElement(
			'svg',
			{
				viewBox: svg_canvas_dimensions,
				className: 'rrui__activity-indicator-circle-container' },
			_react2.default.createElement('circle', {
				className: 'rrui__activity-indicator-circle',
				cx: svg_circe_center,
				cy: svg_circe_center,
				r: radius,
				fill: 'none',
				strokeWidth: radius * 0.125,
				strokeMiterlimit: svg_stroke_miter_limit })
		)
	);
}

ActivityIndicator.propTypes = {
	// CSS class
	className: _propTypes2.default.string,

	// CSS style object
	style: _propTypes2.default.object
};
