// react-responsive-ui/commonjs/PageAndMenu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var PageAndMenu = function (_Component) {
	_inherits(PageAndMenu, _Component);

	// state =
	// {
	// 	show_menu  : false,
	// 	menu_width : 0,
	// 	page_moved_aside : false
	// }

	function PageAndMenu() {
		_classCallCheck(this, PageAndMenu);

		var _this = _possibleConstructorReturn(this, (PageAndMenu.__proto__ || Object.getPrototypeOf(PageAndMenu)).call(this));

		_this.hide_menu_on_mouse_down = _this.hide_menu_on_mouse_down.bind(_this);
		// this.toggle_menu        = this.toggle_menu.bind(this)
		// this.update_menu_width  = this.update_menu_width.bind(this)
		return _this;
	}

	_createClass(PageAndMenu, [{
		key: 'render',
		value: function render() {
			var _props = this.props,
			    children = _props.children,
			    rest = _objectWithoutProperties(_props, ['children']);

			return _react2.default.createElement(
				'div',
				_extends({
					onTouchStart: this.hide_menu_on_mouse_down,
					onMouseDown: this.hide_menu_on_mouse_down
				}, rest),
				children
			);
		}
	}, {
		key: 'hide_menu_on_mouse_down',
		value: function hide_menu_on_mouse_down(event) {
			if (!this.menu || !this.menu_button) {
				return;
			}

			// Hide the menu only if clicked outside
			if (this.menu.element().contains(event.target) || this.menu_button.element().contains(event.target)) {
				return;
			}

			this.menu.hide();
		}

		// toggle_menu()
		// {
		// 	if (!this.state.show_menu)
		// 	{
		// 		return this.setState({ show_menu: !this.state.show_menu, page_moved_aside: !this.state.page_moved_aside })
		// 	}
		//
		// 	this.setState({ show_menu: !this.state.show_menu }, () =>
		// 	{
		// 		// Requires a corresponding `clearTimeout()`` in `componentWillUnmount()``
		// 		setTimeout(() =>
		// 		{
		// 			this.setState({ page_moved_aside: this.state.show_menu })
		// 		},
		// 		menu_transition_duration)
		// 	})
		// }

		// update_menu_width(width)
		// {
		// 	this.setState({ menu_width: width })
		// }

	}, {
		key: 'getChildContext',
		value: function getChildContext() {
			var _this2 = this;

			var context = {
				react_responsive_ui_menu: {
					toggle: function toggle() {
						return _this2.menu.toggle();
					},

					register: function register(menu) {
						if (_this2.menu) {
							throw new Error('There already is a menu registered for this page');
						}

						_this2.menu = menu;

						return function () {
							return _this2.menu = undefined;
						};
					},

					register_menu_button: function register_menu_button(menu_button) {
						if (_this2.menu_button) {
							throw new Error('There already is a menu button registered for this page');
						}

						_this2.menu_button = menu_button;

						return function () {
							return _this2.menu_button = undefined;
						};
					}
				}
			};

			return context;
		}
	}]);

	return PageAndMenu;
}(_react.Component);

PageAndMenu.childContextTypes = {
	react_responsive_ui_menu: _propTypes2.default.shape({
		toggle: _propTypes2.default.func.isRequired,
		register: _propTypes2.default.func.isRequired
	}).isRequired };
exports.default = PageAndMenu;
