// react-responsive-ui/commonjs/SegmentedControl.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _dom = require('./utility/dom');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Segmented_control = function (_PureComponent) {
	_inherits(Segmented_control, _PureComponent);

	function Segmented_control() {
		var _ref;

		var _temp, _this, _ret;

		_classCallCheck(this, Segmented_control);

		for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
			args[_key] = arguments[_key];
		}

		return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Segmented_control.__proto__ || Object.getPrototypeOf(Segmented_control)).call.apply(_ref, [this].concat(args))), _this), _this.state = {}, _this.storeContainerComponent = function (_) {
			return _this.container = _;
		}, _this.onKeyDown = function (event) {
			var onKeyDown = _this.props.onKeyDown;


			if (onKeyDown) {
				onKeyDown(event);
			}

			if ((0, _dom.submitFormOnCtrlEnter)(event, _this.container)) {
				return;
			}

			if (event.ctrlKey || event.altKey || event.shiftKey || event.metaKey) {
				return;
			}

			var _this$props = _this.props,
			    onChange = _this$props.onChange,
			    options = _this$props.options;


			switch (event.keyCode) {
				// Select the previous option (if present) on left arrow
				case 37:
					event.preventDefault();

					var previous_option_index = _this.previous_option_index();

					if (previous_option_index !== undefined) {
						// Focus the option <button/> so that
						// a subsequent "Spacebar" keyDown
						// doesn't select the previously selected
						// option (e.g. the first one)
						_this['button_' + previous_option_index].focus();

						// Change the `value`
						return onChange(options[previous_option_index].value);
					}

					return;

				// Select the next option (if present) on right arrow
				case 39:
					event.preventDefault();

					var next_option_index = _this.next_option_index();

					if (next_option_index !== undefined) {
						// Focus the option <button/> so that
						// a subsequent "Spacebar" keyDown
						// doesn't select the previously selected
						// option (e.g. the first one)
						_this['button_' + next_option_index].focus();

						// Change the `value`
						return onChange(options[next_option_index].value);
					}

					return;
			}
		}, _temp), _possibleConstructorReturn(_this, _ret);
	}

	_createClass(Segmented_control, [{
		key: 'componentDidMount',


		// Client side rendering, javascript is enabled
		value: function componentDidMount() {
			var fallback = this.props.fallback;


			if (fallback) {
				this.setState({ javascript: true });
			}
		}
	}, {
		key: 'render',
		value: function render() {
			var _this2 = this;

			var _props = this.props,
			    disabled = _props.disabled,
			    options = _props.options,
			    fallback = _props.fallback,
			    indicateInvalid = _props.indicateInvalid,
			    error = _props.error,
			    style = _props.style,
			    className = _props.className;


			return _react2.default.createElement(
				'div',
				{
					ref: this.storeContainerComponent,
					onKeyDown: this.onKeyDown,
					className: (0, _classnames2.default)('rrui__segmented-control', {
						'rrui__rich': fallback,
						'rrui__segmented-control--disabled': disabled
					}, className),
					style: style },
				_react2.default.createElement(
					'div',
					{ className: 'rrui__input' },
					options.map(function (option, index) {
						return _this2.render_button(option, index);
					})
				),
				fallback && !this.state.javascript && this.render_static(),
				indicateInvalid && error && _react2.default.createElement(
					'div',
					{ className: 'rrui__input-error' },
					error
				)
			);
		}
	}, {
		key: 'render_button',
		value: function render_button(option, index) {
			var _this3 = this;

			var _props2 = this.props,
			    options = _props2.options,
			    value = _props2.value,
			    disabled = _props2.disabled;


			var selected = value === option.value;

			var first = index === 0;
			var last = index === options.length - 1;

			return _react2.default.createElement(
				'button',
				{
					key: option.value,
					ref: function ref(_ref2) {
						return _this3['button_' + index] = _ref2;
					},
					type: 'button',
					tabIndex: index === 0 ? undefined : '-1',
					disabled: disabled,
					onClick: this.chooser(option.value),
					className: (0, _classnames2.default)('rrui__button-reset', 'rrui__segmented-control__option', {
						'rrui__segmented-control__option--selected': selected,
						// CSS selector performance optimization
						'rrui__segmented-control__option--disabled': disabled,
						// Ordering
						'rrui__segmented-control__option--first': first,
						'rrui__segmented-control__option--last': last,
						'rrui__segmented-control__option--middle': !first && !last
					}) },
				option.label
			);
		}

		// supports disabled javascript

	}, {
		key: 'render_static',
		value: function render_static() {
			var _this4 = this;

			var options = this.props.options;


			return _react2.default.createElement(
				'div',
				{ className: 'rrui__rich__fallback' },
				options.map(function (option, index) {
					return _this4.render_static_option(option, index);
				})
			);
		}
	}, {
		key: 'render_static_option',
		value: function render_static_option(option, index) {
			var _props3 = this.props,
			    options = _props3.options,
			    name = _props3.name,
			    value = _props3.value;


			var first = index === 0;
			var last = index === options.length - 1;

			return _react2.default.createElement(
				'span',
				{
					key: option.value,
					className: (0, _classnames2.default)('rrui__button-reset', 'rrui__segmented-control__option', {
						'rrui__segmented-control__option--first': first,
						'rrui__segmented-control__option--last': last,
						'rrui__segmented-control__option--middle': !first && !last
					}) },
				_react2.default.createElement('input', {
					type: 'radio',
					name: name,
					checked: value === option.value }),
				option.label
			);
		}
	}, {
		key: 'chooser',
		value: function chooser(value) {
			var _this5 = this;

			return function (event) {
				var _props4 = _this5.props,
				    disabled = _props4.disabled,
				    onChange = _props4.onChange,
				    previous_value = _props4.value;


				if (disabled) {
					return;
				}

				// Call `onChange` only if `value` did actually change
				if (value !== previous_value) {
					onChange(value);
				}
			};
		}
	}, {
		key: 'focus',
		value: function focus() {
			_reactDom2.default.findDOMNode(this.button_0).focus();
		}
	}, {
		key: 'previous_option_index',


		// Get the previous option index (relative to the currently selected option)
		value: function previous_option_index() {
			var _props5 = this.props,
			    options = _props5.options,
			    value = _props5.value;


			var i = 0;
			while (i < options.length) {
				if (options[i].value === value) {
					if (i - 1 >= 0) {
						return i - 1;
					}
				}
				i++;
			}
		}

		// Get the next option index (relative to the currently selected option)

	}, {
		key: 'next_option_index',
		value: function next_option_index() {
			var _props6 = this.props,
			    options = _props6.options,
			    value = _props6.value;


			var i = 0;
			while (i < options.length) {
				if (options[i].value === value) {
					if (i + 1 < options.length) {
						return i + 1;
					}
				}
				i++;
			}
		}
	}]);

	return Segmented_control;
}(_react.PureComponent);

Segmented_control.propTypes = {
	// A list of selectable options
	options: _propTypes2.default.arrayOf(_propTypes2.default.shape({
		// Option value
		value: _propTypes2.default.string.isRequired,
		// Option label
		label: _propTypes2.default.string.isRequired
	})).isRequired,

	// HTML form input `name` attribute
	name: _propTypes2.default.string,

	// If `true` then will be disabled
	disabled: _propTypes2.default.bool,

	// The selected option value
	value: _propTypes2.default.any,

	// Is called when an option is selected
	onChange: _propTypes2.default.func.isRequired,

	// (exotic use case)
	// Falls back to a plain HTML input
	// when javascript is disabled (e.g. Tor)
	fallback: _propTypes2.default.bool.isRequired,

	// CSS class
	className: _propTypes2.default.string,

	// CSS style object
	style: _propTypes2.default.object
};
Segmented_control.defaultProps = {
	fallback: false };
exports.default = Segmented_control;
